### ISSUE TITLE
`active` and `paused` props don't work

### ISSUE THREAD

**Issue by Dan503** [focus-trap/focus-trap-react#239]
I have smoothly animating dialogs that fade in and fade out.

All the examples assume that the dialogs have no intro or outro animations.

I've forked the repository and converted the "demo-defaults" example into a test case example of what I need FocusTrap to be able to do:
https://github.com/Dan503/focus-trap-react/blob/master/demo/js/demo-defaults.js#L36

There are `active` and `paused` props according to the index.d.ts file but when I try to use them, I get this error:

```
Uncaught Error: Your focus-trap must have at least one container with at least one tabbable node in it at all times
```

Basically this is what I'm asking for (in pseudo code terms):

````
if ( `active` prop equals `true` ) {
    activate the focus-trap
}
else if ( `active` prop equals `false` ) {
    deactivate the focus-trap
}

if ( `paused` prop equals `true` ) {
    pause the focus-trap
}
else if ( `paused` prop equals `false` ) {
    resume the focus-trap
}
````

**Comment by stefcameron** [comment]
Hmmm, well, the __special element__ demo does feature a trap that is activated and deactivated while remaining mounted, successfully using the `active` prop to activate and deactivate it in the same way your pseudo code indicates it should work.

Here is a modified version of the same demo where I've added a un/pause toggle button that can be used to pause and unpause the trap in the same way your pseudo code would like to do it:

![Kapture 2021-01-09 at 12 28 03](https://user-images.githubusercontent.com/2855350/104105797-60e26380-5276-11eb-8328-5b35155becf0.gif)

```javascript
// file: /demo/js/demo-special-element.js

const React = require('react');
const ReactDOM = require('react-dom');
const FocusTrap = require('../../dist/focus-trap-react');

const container = document.getElementById('demo-special-element');

class DemoSpecialElement extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      activeTrap: false,
      pausedTrap: false,
      passThruMsg: '',
    };

    this.mountTrap = this.mountTrap.bind(this);
    this.pauseTrap = this.pauseTrap.bind(this);
    this.unmountTrap = this.unmountTrap.bind(this);
    this.updatePassThruMsg = this.updatePassThruMsg.bind(this);
  }

  mountTrap() {
    this.setState({ activeTrap: true, passThruMsg: '' });
  }

  pauseTrap() {
    this.setState({ pausedTrap: !this.state.pausedTrap });
  }

  unmountTrap() {
    this.setState({ activeTrap: false });
  }

  updatePassThruMsg() {
    this.setState({ passThruMsg: 'Clicked!' });
  }

  render() {
    let trapClass = 'trap';
    if (this.state.activeTrap) {
      trapClass += ' is-active';
    }

    return (
      <div>
        <p>
          <button
            id="demo-special-element-activate"
            onClick={this.mountTrap}
            aria-label="activate trap for 'special element' demo"
          >
            activate trap
          </button>
          <button
            id="demo-special-element-pause"
            onClick={this.pauseTrap}
            aria-label="pause trap for 'special element' demo"
          >
            {this.state.pausedTrap ? 'unpause' : 'pause'} trap
          </button>
          <button onClick={this.updatePassThruMsg}>pass thru click</button>
          <span>{this.state.passThruMsg}</span>
        </p>
        <FocusTrap
          active={this.state.activeTrap}
          paused={this.state.pausedTrap}
          focusTrapOptions={{
            onDeactivate: this.unmountTrap,
            clickOutsideDeactivates: false,
            allowOutsideClick(event) {
              return event.target.id === 'demo-special-element-activate' ||
                  event.target.id === 'demo-special-element-pause';
            },
            returnFocusOnDeactivate: true,
          }}
        >
          <section
            id="focus-trap-three"
            style={this.state.activeTrap ? null : { background: '#eee' }}
            data-whatever="nothing"
            className={trapClass}
          >
            <p>
              Here is a focus trap <a href="#">with</a> <a href="#">some</a>{' '}
              <a href="#">focusable</a> parts.
            </p>
            <p>
              <button
                onClick={this.unmountTrap}
                aria-label="deactivate trap for 'special element' demo"
              >
                deactivate trap
              </button>
            </p>
          </section>
        </FocusTrap>
      </div>
    );
  }
}

ReactDOM.render(<DemoSpecialElement />, container);
```

Now why are you getting those errors? My best guess (and thank you for the fork and modified demo -- that's a big help! 😄 ) is that you're hitting this condition in tabbable: https://github.com/focus-trap/tabbable/blob/master/src/index.js#L144-L146

That being because of timing challenges with what you're trying to do: https://github.com/Dan503/focus-trap-react/blob/master/demo/js/demo-defaults.js#L18

If you have any ideas on how you would enhance focus-trap-react to support this kind of delay, I'm happy to look at a PR, but it feels a bit like a corner case. I'm sure others have faced similar issues, but I'm sure it's common enough. Anyway, if you want to propose something, I'm happy to consider it, as long as it's generic enough and doesn't affect the usual "synchronicity" most consumers are used to with this library.

Did you try adding another state flag that controls the animation, such that when the animation is complete and the node is visible, then the trap gets activated? It's not like users can interact with the content being animated while it's being animated into view anyway, so why not activate immediately once the animation is done?

Maybe you could use a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver), [like this](https://github.com/Mirantis/lens-extension-cc/blob/master/src/page.tsx#L39), to efficiently detect when the class names change to indicate it's visible.

**Comment by Dan503** [comment]
Hmm... I wonder if adding `activationDelay` and `pauseDelay` options to the Focus-Trap options object would be a viable solution.

**Comment by stefcameron** [comment]
> Hmm... I wonder if adding `activationDelay` and `pauseDelay` options to the Focus-Trap options object would be a viable solution.

Maybe that would work, but you'd have to keep that in sync with whatever other timings you've got going on. Not sure how reliable that would be, though, since timing of events in the browser is not guaranteed to be super precise all the time. Lots of things could trip it up, like other processes causing a slight slow down, for example. It might be better to have a more definite trigger for the delayed activation, such as watching for a change in class names or something.

> Although since this is something that is more of a focus-trap-react issue, it might be better making them props on the FocusTrap component instead.

Most likely. This doesn't feel like it would be React-specific.

**Comment by Dan503** [comment]
> It might be better to have a more definite trigger for the delayed activation, such as watching for a change in class names or something.

I don't think that is an option since the timing is done on the CSS side by the rendering engine. The class gets applied instantly, but the visibility doesn't actually change for another 0.5 seconds when transitioning from visible to hidden.

**Comment by Dan503** [comment]
> Not sure how reliable that would be, though, since timing of events in the browser is not guaranteed to be super precise all the time.

I don't think it needs to be super precise for this use case anyway. If setting the delay to 0.5s sometimes causes something to go wrong, you can set `activationDelay` to 0.6s instead. The user won't notice an extra 0.1s delay after the animation ends.

**Comment by stefcameron** [comment]
@Dan503 Hey, I'm wondering, did you end-up solving your problem in another way?

**Comment by Dan503** [comment]
No I haven't solved it yet.

I started working on a PR to fix the issue but I got distracted with other stuff.

I also ran into some compile issues and didn't want to create a PR that potentially broke your build.

**Comment by stefcameron** [comment]
OK, just checking. Thanks for starting a PR! LMK if I can help sort out those compile issues whenever you get back to it.

**Comment by Dan503** [comment]
Pull request created for focus-trap:
https://github.com/focus-trap/focus-trap/pull/407

**Comment by stefcameron** [comment]
This will be fixed when https://github.com/focus-trap/focus-trap/pull/407 is taken into `focus-trap-react` with `focus-trap@6.5.0`

**Comment by stefcameron** [comment]
Published in v8.5.1, which bumps to `focus-trap@6.5.0`, which provides the new features needed to handle this.

**Comment by stefcameron** [comment]
This was ultimately fixed with #394 that was published in v8.6.0.
